pkg/resource/graph/dependency_graph_test.go

```diff
@@ -75,841 +82,1680 @@ func TestDependencyGraph(t *testing.T) {
+		})
+
+		t.Run("e1", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := []*resource.State{e2, e3, e4, e5}
+
+			// Act.
+			actual := dg.OnlyDependsOn(e1)
+
+			// Assert.
+			assertSameStates(t, expected, actual)
+		})
+
+		// The point of OnlyDependsOn is to avoid finding resources which might depend on multiple resources with the same
+		// URN (but different IDs, for instance). This happens particularly in the case where a provider might be being
+		// upgraded, for example. The following test cases bulk out the suite for OnlyDependsOn to really exercise this
+		// behaviour.
+
+		providerF1, providerF1Ref := makeProvider("pkg", "providerF", "1")
+		providerF2, providerF2Ref := makeProvider("pkg", "providerF", "2")
+		providerF3, providerF3Ref := makeProvider("pkg", "providerF", "3")
+
+		fx1 := &resource.State{URN: "fx", Provider: providerF1Ref}
+		fx2 := &resource.State{URN: "fx", Provider: providerF2Ref}
+
+		fy := &resource.State{URN: "fy", Provider: providerF3Ref, Dependencies: []resource.URN{fx1.URN}}
+		fz := &resource.State{
+			URN:      "fz",
+			Provider: providerF3Ref,
+			PropertyDependencies: map[resource.PropertyKey][]resource.URN{
+				"fzProp1": {fx1.URN},
 			},
 		}
 
-		for _, c := range cases {
-			c := c
-			t.Run(c.name, func(t *testing.T) {
-				t.Parallel()
+		fw := &resource.State{
+			URN:         "fw",
+			Provider:    providerF3Ref,
+			DeletedWith: fx1.URN,
+		}
+
+		fu := &resource.State{URN: "fu", Provider: providerF3Ref}
+		fv := &resource.State{
+			URN:         "fv",
+			Provider:    providerF3Ref,
+			DeletedWith: fu.URN,
+		}
+
+		dgOnly := NewDependencyGraph([]*resource.State{
+			providerF1,
+			providerF2,
+			providerF3,
+			fx1,
+			fx2,
+			fy,
+			fz,
+			fw,
+			fu,
+			fv,
+		})
+
+		t.Run("providerF1", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := []*resource.State{fx1}
+
+			// Act.
+			actual := dgOnly.OnlyDependsOn(providerF1)
+
+			// Assert.
+			assertSameStates(t, expected, actual)
+		})
+
+		t.Run("providerF2", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := []*resource.State{fx2}
+
+			// Act.
+			actual := dgOnly.OnlyDependsOn(providerF2)
+
+			// Assert.
+			assertSameStates(t, expected, actual)
+		})
+
+		t.Run("providerF3", func(t *testing.T) {
+			t.Parallel()
+
+			// Arrange.
+			expected := []*resource.State{
+				fy,
+				fz,
+
+				// BUG: fw actually depends on fx1, which depends on providerF1, so should not be returned.
+				fw,
+				fu,
+				fv,
+			}
+
+			// Act.
+			actual := dgOnly.OnlyDependsOn(providerF3)
+
+			// Assert.
+			assertSameStates(t, expected, actual)
+		})
+
+		t.Run("fx1", func(t *testing.T) {
+			t.Parallel()
+
+			// Act.
+			actual := dgOnly.OnlyDependsOn(fx1)
+
+			// Assert.
+			assertSameStates(t, nil, actual)
+		})
+
+		t.Run("fx2", func(t *testing.T) {
+			t.Parallel()
+
+			// Act.
+			actual := dgOnly.OnlyDependsOn(fx2)
+
+			// Assert.
+			assertSameStates(t, nil, actual)
```

Comment:
This is technically correct because object identity, but the current docs don't make this clear and I stared at this for a long time. This is another thing I intend to address in a future PR.